File: ios/CareLog/CareLog/History/ViewModels/HistoryViewModel.swift
import Foundation
import SwiftUI
import Combine

/// ViewModel for history screen.
@MainActor
final class HistoryViewModel: ObservableObject {
    // MARK: - Published Properties

    @Published var entries: [HistoryEntry] = []
    @Published var selectedVitalType: DashboardItem?
    @Published var showDateFilter: Bool = false
    @Published var startDate: Date?
    @Published var endDate: Date?
    @Published var isLoading: Bool = false

    // MARK: - Dependencies

    private let authService: AuthService
    private let localFHIRRepository: LocalFHIRRepository
    private var cancellables = Set<AnyCancellable>()

    // MARK: - Initialization

    init(
        authService: AuthService = .shared,
        localFHIRRepository: LocalFHIRRepository = .shared
    ) {
        self.authService = authService
        self.localFHIRRepository = localFHIRRepository

        Task {
            await loadHistory()
        }

        // Reload when filters change
        $startDate
            .dropFirst()
            .sink { [weak self] _ in
                Task { await self?.loadHistory() }
            }
            .store(in: &cancellables)

        $endDate
            .dropFirst()
            .sink { [weak self] _ in
                Task { await self?.loadHistory() }
            }
            .store(in: &cancellables)
    }

    // MARK: - Actions

    func selectVitalType(_ type: DashboardItem?) {
        selectedVitalType = type
        Task {
            await loadHistory()
        }
    }

    func toggleDateFilter() {
        showDateFilter.toggle()
    }

    func clearDateFilter() {
        startDate = nil
        endDate = nil
    }

    // MARK: - Data Loading

    private func loadHistory() async {
        isLoading = true

        do {
            guard let user = await authService.getCurrentUser(),
                  let patientId = user.linkedPatientId else {
                entries = []
                isLoading = false
                return
            }

            // Get observations based on filter
            let observationType = selectedVitalType?.toObservationType()
            let observations: [FHIRObservation]

            if let type = observationType {
                observations = try localFHIRRepository.getObservationsByType(patientId, type: type)
            } else {
                observations = try localFHIRRepository.getObservationsForPatient(patientId)
            }

            // Filter by date range if set
            let filteredObservations = observations.filter { obs in
                let obsDate = obs.effectiveDateTime

                let afterStart = startDate.map { obsDate >= $0 } ?? true
                let beforeEnd = endDate.map { obsDate <= $0 } ?? true

                return afterStart && beforeEnd
            }

            // Convert to history entries
            entries = filteredObservations.map { obs in
                HistoryEntry(
                    id: obs.id ?? UUID().uuidString,
                    vitalType: obs.type.toDashboardItem(),
                    displayValue: formatDisplayValue(obs),
                    timestamp: obs.effectiveDateTime,
                    performerName: obs.performerId,
                    syncStatus: .synced
                )
            }.sorted { $0.timestamp > $1.timestamp }

            isLoading = false

        } catch {
            entries = []
            isLoading = false
        }
    }

    private func formatDisplayValue(_ obs: FHIRObservation) -> String {
        switch obs.type {
        case .bloodPressure:
            let systolic = obs.components?.first(where: { $0.type == .systolicBP })?.value
            let diastolic = obs.components?.first(where: { $0.type == .diastolicBP })?.value
            if let sys = systolic, let dia = diastolic {
                return "\(Int(sys))/\(Int(dia)) mmHg"
            }
            return "-- mmHg"

        case .bloodGlucose:
            if let value = obs.value {
                return "\(Int(value)) mg/dL"
            }
            return "-- mg/dL"

        case .bodyTemperature:
            if let celsius = obs.value {
                let fahrenheit = celsius * 9 / 5 + 32
                return String(format: "%.1f\u{00B0}F", fahrenheit)
            }
            return "--\u{00B0}F"

        case .bodyWeight:
            if let value = obs.value {
                return String(format: "%.1f kg", value)
            }
            return "-- kg"

        case .heartRate:
            if let value = obs.value {
                return "\(Int(value)) bpm"
            }
            return "-- bpm"

        case .oxygenSaturation:
            if let value = obs.value {
                return "\(Int(value))%"
            }
            return "--%"

        default:
            return "\(obs.value ?? 0) \(obs.unit ?? "")"
        }
    }
}

// MARK: - History Entry Model

struct HistoryEntry: Identifiable {
    let id: String
    let vitalType: DashboardItem
    let displayValue: String
    let timestamp: Date
    let performerName: String?
    let syncStatus: SyncStatus
}

// MARK: - Extensions

extension DashboardItem {
    func toObservationType() -> ObservationType? {
        switch self {
        case .bloodPressure: return .bloodPressure
        case .glucose: return .bloodGlucose
        case .temperature: return .bodyTemperature
        case .weight: return .bodyWeight
        case .pulse: return .heartRate
        case .spO2: return .oxygenSaturation
        default: return nil
        }
    }
}

extension ObservationType {
    func toDashboardItem() -> DashboardItem {
        switch self {
        case .bloodPressure, .systolicBP, .diastolicBP: return .bloodPressure
        case .bloodGlucose: return .glucose
        case .bodyTemperature: return .temperature
        case .bodyWeight: return .weight
        case .heartRate: return .pulse
        case .oxygenSaturation: return .spO2
        default: return .bloodPressure
        }
    }
}

File: android/app/src/main/java/com/carelog/ui/history/HistoryViewModel.kt
package com.carelog.ui.history

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.carelog.auth.AuthRepository
import com.carelog.fhir.client.ObservationType
import com.carelog.fhir.local.entities.SyncStatus
import com.carelog.fhir.repository.LocalFhirRepository
import com.carelog.ui.dashboard.VitalType
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.first
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import java.time.Instant
import java.time.LocalDate
import java.time.ZoneId
import javax.inject.Inject

/**
 * ViewModel for history screen.
 */
@HiltViewModel
class HistoryViewModel @Inject constructor(
    private val authRepository: AuthRepository,
    private val localFhirRepository: LocalFhirRepository
) : ViewModel() {

    private val _uiState = MutableStateFlow(HistoryUiState())
    val uiState: StateFlow<HistoryUiState> = _uiState.asStateFlow()

    init {
        loadHistory()
    }

    fun selectVitalType(type: VitalType?) {
        _uiState.update { it.copy(selectedVitalType = type) }
        loadHistory()
    }

    fun toggleDateFilter() {
        _uiState.update { it.copy(showDateFilter = !it.showDateFilter) }
    }

    fun setStartDate(date: LocalDate) {
        _uiState.update { it.copy(startDate = date) }
        loadHistory()
    }

    fun setEndDate(date: LocalDate) {
        _uiState.update { it.copy(endDate = date) }
        loadHistory()
    }

    fun clearDateFilter() {
        _uiState.update { it.copy(startDate = null, endDate = null) }
        loadHistory()
    }

    private fun loadHistory() {
        viewModelScope.launch {
            _uiState.update { it.copy(isLoading = true) }

            try {
                val user = authRepository.getCurrentUser()
                val patientId = user?.linkedPatientId

                if (patientId == null) {
                    _uiState.update { it.copy(isLoading = false, entries = emptyList()) }
                    return@launch
                }

                val state = _uiState.value

                // Get observations based on filter
                val observationType = state.selectedVitalType?.toObservationType()
                val observationsFlow = if (observationType != null) {
                    localFhirRepository.getObservationsByType(patientId, observationType)
                } else {
                    localFhirRepository.getObservationsForPatient(patientId)
                }

                // Collect the first emission from the flow
                val observations = observationsFlow.first()

                // Filter by date range if set
                val filteredObservations = observations.filter { obs: com.carelog.fhir.models.FhirObservation ->
                    val obsDate = Instant.parse(obs.effectiveDateTime)
                        .atZone(ZoneId.systemDefault())
                        .toLocalDate()

                    val afterStart = state.startDate?.let { obsDate >= it } ?: true
                    val beforeEnd = state.endDate?.let { obsDate <= it } ?: true

                    afterStart && beforeEnd
                }

                // Convert to history entries
                val entries = filteredObservations.map { obs: com.carelog.fhir.models.FhirObservation ->
                    HistoryEntry(
                        id = obs.id ?: "",
                        vitalType = obs.type.toVitalType(),
                        displayValue = formatDisplayValue(obs.type, obs.value, obs.unit, obs.components),
                        timestamp = Instant.parse(obs.effectiveDateTime),
                        performerName = null,
                        syncStatus = SyncStatus.SYNCED
                    )
                }.sortedByDescending { it.timestamp }

                _uiState.update { it.copy(isLoading = false, entries = entries) }

            } catch (e: Exception) {
                _uiState.update { it.copy(isLoading = false, entries = emptyList()) }
            }
        }
    }

    private fun formatDisplayValue(
        type: ObservationType,
        value: Double?,
        unit: String?,
        components: List<com.carelog.fhir.models.ObservationComponent>?
    ): String {
        return when (type) {
            ObservationType.BLOOD_PRESSURE -> {
                val systolic = components?.find { it.type == ObservationType.SYSTOLIC_BP }?.value?.toInt()
                val diastolic = components?.find { it.type == ObservationType.DIASTOLIC_BP }?.value?.toInt()
                if (systolic != null && diastolic != null) {
                    "$systolic/$diastolic mmHg"
                } else {
                    "-- mmHg"
                }
            }
            ObservationType.BLOOD_GLUCOSE -> "${value?.toInt() ?: "--"} mg/dL"
            ObservationType.BODY_TEMPERATURE -> {
                val fahrenheit = value?.let { it * 9 / 5 + 32 }
                String.format("%.1f°F", fahrenheit ?: 0.0)
            }
            ObservationType.BODY_WEIGHT -> String.format("%.1f kg", value ?: 0.0)
            ObservationType.HEART_RATE -> "${value?.toInt() ?: "--"} bpm"
            ObservationType.OXYGEN_SATURATION -> "${value?.toInt() ?: "--"}%"
            else -> "${value ?: "--"} ${unit ?: ""}"
        }
    }
}

/**
 * Extension to convert VitalType to ObservationType.
 */
private fun VitalType.toObservationType(): ObservationType? {
    return when (this) {
        VitalType.BLOOD_PRESSURE -> ObservationType.BLOOD_PRESSURE
        VitalType.GLUCOSE -> ObservationType.BLOOD_GLUCOSE
        VitalType.TEMPERATURE -> ObservationType.BODY_TEMPERATURE
        VitalType.WEIGHT -> ObservationType.BODY_WEIGHT
        VitalType.PULSE -> ObservationType.HEART_RATE
        VitalType.SPO2 -> ObservationType.OXYGEN_SATURATION
        else -> null
    }
}

/**
 * Extension to convert ObservationType to VitalType.
 */
private fun ObservationType.toVitalType(): VitalType {
    return when (this) {
        ObservationType.BLOOD_PRESSURE -> VitalType.BLOOD_PRESSURE
        ObservationType.SYSTOLIC_BP -> VitalType.BLOOD_PRESSURE
        ObservationType.DIASTOLIC_BP -> VitalType.BLOOD_PRESSURE
        ObservationType.BLOOD_GLUCOSE -> VitalType.GLUCOSE
        ObservationType.BODY_TEMPERATURE -> VitalType.TEMPERATURE
        ObservationType.BODY_WEIGHT -> VitalType.WEIGHT
        ObservationType.HEART_RATE -> VitalType.PULSE
        ObservationType.OXYGEN_SATURATION -> VitalType.SPO2
    }
}

/**
 * UI state for history screen.
 */
data class HistoryUiState(
    val entries: List<HistoryEntry> = emptyList(),
    val selectedVitalType: VitalType? = null,
    val showDateFilter: Boolean = false,
    val startDate: LocalDate? = null,
    val endDate: LocalDate? = null,
    val isLoading: Boolean = false
)

/**
 * Single history entry model.
 */
data class HistoryEntry(
    val id: String,
    val vitalType: VitalType,
    val displayValue: String,
    val timestamp: Instant,
    val performerName: String?,
    val syncStatus: SyncStatus
)
